data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/export/ExportScheduler.java

```diff
@@ -32,17 +39,183 @@ public class ExportScheduler implements Runnable {
+                    break;
+                }
+            }
+        }
+        LOG.warn("Export scheduler interrupted, looks like shutdown has triggered");
+        executor.shutdownNow();
+    }
+
+    private String getOrCreateExportTaskId(ExportPartition exportPartition) {
+        ExportProgressState progressState = exportPartition.getProgressState().get();
+
+        if (progressState.getExportTaskId() != null) {
+            LOG.info("Export task has already created for db {}", exportPartition.getDbIdentifier());
+            return progressState.getExportTaskId();
+        }
+
+        LOG.info("Creating a new snapshot for db {}", exportPartition.getDbIdentifier());
+        SnapshotInfo snapshotInfo = snapshotManager.createSnapshot(exportPartition.getDbIdentifier());
+        if (snapshotInfo != null) {
+            LOG.info("Snapshot id is {}", snapshotInfo.getSnapshotId());
+            progressState.setSnapshotId(snapshotInfo.getSnapshotId());
+            sourceCoordinator.saveProgressStateForPartition(exportPartition, null);
+        } else {
+            LOG.error("The snapshot failed to create, it will be retried");
+            closeExportPartitionWithError(exportPartition);
+            return null;
+        }
+
+        final String snapshotId = snapshotInfo.getSnapshotId();
+        try {
+            checkSnapshotStatus(snapshotId);
+        } catch (Exception e) {
+            LOG.warn("Check snapshot status for {} failed", snapshotId, e);
+            sourceCoordinator.giveUpPartition(exportPartition);
+        }
+
+        LOG.info("Creating an export task for db {} from snapshot {}", exportPartition.getDbIdentifier(), snapshotId);
+        String exportTaskId = exportTaskManager.startExportTask(
+                snapshotInfo.getSnapshotArn(), progressState.getIamRoleArn(), progressState.getBucket(),
+                progressState.getPrefix(), progressState.getKmsKeyId(), progressState.getTables());
+
+        if (exportTaskId != null) {
+            LOG.info("Export task id is {}", exportTaskId);
+            progressState.setExportTaskId(exportTaskId);
+            sourceCoordinator.saveProgressStateForPartition(exportPartition, null);
+        } else {
+            LOG.error("The export task failed to create, it will be retried");
+            closeExportPartitionWithError(exportPartition);
+        }
+
+        return exportTaskId;
+    }
+
+    private void closeExportPartitionWithError(ExportPartition exportPartition) {
+        ExportProgressState exportProgressState = exportPartition.getProgressState().get();
+        // Clear current task id, so that a new export can be submitted.
+        exportProgressState.setExportTaskId(null);
+        sourceCoordinator.closePartition(exportPartition, DEFAULT_CLOSE_DURATION, DEFAULT_MAX_CLOSE_COUNT);
+    }
+
+    private String checkSnapshotStatus(String snapshotId) {
+        LOG.debug("Start checking status of snapshot {}", snapshotId);
+        while (true) {
```

Comment:
See my comment below about infinite loops.

Reply:
Added a timeout to the loop.